src/adapter/src/subscribe.rs

```diff
@@ -113,7 +219,18 @@ impl ActiveSubscribe {
                                     packer.push(Datum::False);
                                 }
 
-                                packer.push(Datum::Int64(diff));
+                                if matches!(self.output, SubscribeOutput::EnvelopeUpsert { .. }) {
+                                    packer.push(match diff {
+                                        -1 => Datum::String("delete"),
```

Comment:
I am afraid there is a bit of a problem with the `delete` state. Consider this case:

```
DROP TABLE IF EXISTS t1 CASCADE;
CREATE TABLE t1 (f1 INTEGER);
INSERT INTO t1 VALUES (1), (2);
COPY (SUBSCRIBE t1 ENVELOPE UPSERT (KEY (f1))) TO STDOUT;
```
and then in another connection:

```
update t1 set f1 = 5;
```

The `COPY` returns the following output:

```
1682670801011	delete	1
1682670801011	delete	2
1682670801011	key violation	5
```

So if I was a customer that was ingesting this output, I would act on the deletes and then, ops, error, and my data is gone :-( . I think it will be safest if this situation either:
 - the SUBSCRIBE errors out completely without emitting a "delete", or
 - emit the "key violation" but do not emit any "deletes"

Reply:
Actually I think the issue here comes from the fact that one is allowed to declare `UPSERT (KEY)` on a key that did not originally exist. This can not happen with sinks, as they have an explicit protection against that. See the end of `kafka-avro-upsert-sinks.td` . Maybe such a protection should be added to `SUBSCRIBE` as well:

With sinks, an error:

```
materialize=> CREATE SINK invalid_sub_key FROM input_keyed_ab INTO KAFKA CONNECTION kafka_conn (TOPIC 'input-sink-102568587') KEY (a) FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn ENVELOPE UPSERT;
ERROR:  Invalid upsert key: (a), valid keys are: (a, b)
```

With the equivalent subscribe:

```
materialize=> copy (subscribe input_keyed_ab ENVELOPE UPSERT (KEY (a))) to stdout;
3	key violation	1	\N
3	upsert	2	2
3	upsert	3	1
3	upsert	4	2
```

To get a kafka source that you can experiment with:

````
cd test/testdrive
./mzcompose down -v ; ./mzcompose run default kafka-avro-upsert-sinks.td
psql -p  `docker port testdrive-materialized-1 | grep '0.0' | grep 6875 | cut -d':' -f2` -h 127.0.0.1 -U materialize
```

Reply:
The first example you have makes sense in a way: when changing the two rows to 5, we see `(k=1, v=(), -1), (k=2, v=(), -1), (k=5, v=(), +1), (k=5, v=(), +1)`. the first two are normal deletes, the key violation is just when we see `(k=5, v=(), +1), (k=5, v=(), +1)`. I agree with you that this behavior is weird but I think it's worth putting it out there and iterating. I don't think we should expose this w/o a flag before figuring out these semantics

Reply:
ok fair enough, please proceed as you see fit.